internal/sourcemap/sourcemap_fetcher.go

```diff
@@ -0,0 +1,107 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package sourcemap
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/go-sourcemap/sourcemap"
+
+	"github.com/elastic/apm-server/internal/logs"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type SourcemapFetcher struct {
+	metadata MetadataFetcher
+	backend  Fetcher
+	logger   *logp.Logger
+}
+
+func NewSourcemapFetcher(metadata MetadataFetcher, backend Fetcher) *SourcemapFetcher {
+	s := &SourcemapFetcher{
+		metadata: metadata,
+		backend:  backend,
+		logger:   logp.NewLogger(logs.Sourcemap),
+	}
+
+	return s
+}
+
+func (s *SourcemapFetcher) Fetch(ctx context.Context, name, version, path string) (*sourcemap.Consumer, error) {
+	original := identifier{name: name, version: version, path: path}
+
+	select {
+	case <-s.metadata.ready():
+		if err := s.metadata.err(); err != nil {
+			return nil, err
+		}
+	case <-ctx.Done():
+		return nil, fmt.Errorf("error waiting for metadata fetcher to be ready: %w", ctx.Err())
+	}
+
+	// the mutex is shared by the update goroutine, we need to release it
+	// as soon as possible to avoid blocking updates.
```

Comment:
Is this a stale comment? I don't see any mutexes

Reply:
It is, the mutex was inside `getID` but that has been moved to another struct